instant-xml/src/parse.rs

```diff
@@ -1,48 +1,142 @@
-            Some(Ok(Token::ElementEnd { end, .. })) => match end {
-                ElementEnd::Open => todo!(),
-                ElementEnd::Close(_, _) => todo!(),
-                ElementEnd::Empty => Ok(()),
-            },
-            Some(Ok(_)) => Err(Error::UnexpectedValue),
-            Some(Err(err)) => Err(err.into()),
-            None => Err(Error::UnexpectedEndOfStream),
+    pub fn peek_next_tag(&mut self) -> Result<Option<XmlRecord>, Error> {
+        let item = match self.iter.peek() {
+            Some(v) => v,
+            None => return Ok(None),
+        };
+
+        println!("peek: {:?}", &item);
+        match item {
+            Ok(Token::ElementStart { prefix, local, .. }) => {
+                let prefix = match prefix.is_empty() {
+                    true => None,
+                    false => Some(prefix.as_str()),
+                };
+
+                Ok(Some(XmlRecord::Open(TagData {
+                    key: local,
+                    attributes: Vec::new(),
+                    default_namespace: None,
+                    namespaces: None,
+                    prefix,
+                })))
+            }
+            Ok(Token::ElementEnd { end, .. }) => {
+                if let ElementEnd::Close(..) = end {
+                    if self.stack.is_empty() {
+                        return Err(Error::UnexpectedEndOfStream);
+                    }
+
+                    return Ok(Some(XmlRecord::Close(self.stack.last().unwrap())));
+                }
+                Err(Error::UnexpectedToken)
+            }
+            Ok(_) => Err(Error::UnexpectedToken),
+            Err(e) => Err(Error::Parse(*e)),
         }
     }
 }
 
-pub trait Parse {
-    fn element_start(self, ns: Option<&str>, tag: &str) -> Result<(), Error>;
-    fn element_end(self, ns: Option<&str>, tag: &str) -> Result<(), Error>;
+impl<'xml> Iterator for XmlParser<'xml> {
+    type Item = Result<XmlRecord<'xml>, Error>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut key: Option<&str> = None;
+        let mut prefix_ret: Option<&str> = None;
+        let mut default_namespace = None;
+        let mut namespaces = HashMap::new();
+        let mut attributes = Vec::new();
+
+        loop {
+            let item = match self.iter.next() {
+                Some(v) => v,
+                None => return None,
+            };
+
+            println!("{:?}", &item);
+            match item {
+                Ok(Token::ElementStart { prefix, local, .. }) => {
+                    key = Some(local.as_str());
+                    prefix_ret = match prefix.is_empty() {
+                        true => None,
+                        false => Some(prefix.as_str()),
+                    };
+                }
+                Ok(Token::ElementEnd { end, .. }) => match end {
+                    ElementEnd::Open => {
+                        self.stack.push(key.unwrap());
+                        println!(
```

Comment:
Should remove all the `println!()` stuff before merging.

Reply:
So time to do this?

Reply:
I believe so, I put a hold on this, in case you have some additional comments.

Reply:
Nah, I think this is in a decent shape, time to move on!

Reply:
done